### DISCUSSION TITLE
[Detector Support]: Hailo8L from Raspberry {io AI kit problem on Home Assistant Rapsberry Pi5 with Dual M.2 Hat

### DISCUSSION THREAD

**Discussion by roblamoreaux** [blakeblackshear/frigate#15008] [Detector Support]
### Describe the problem you are having

I just got a Raspberry Pi AI Kit and put the M.2 card on my dual M.2 hat with my SSD. When I enable the detector it crashes opening the Hailo8L

### Version

0.15.0-beta1

### Frigate config file

```yaml
mqtt:
  host: 192.168.111.20
  user: MQTTH
  password: Password123

go2rtc:
  streams:
    drivewaycars_camera:
      - rtsp://admin:password@192.168.111.23:8554/Streaming/Channels/101
    drivewaycars_camera_sub:
      - rtsp://admin:password@192.168.111.23:8554/Streaming/Channels/102
    pool_camera:
      - onvif://admin:password1072@192.168.111.32:8000?subtype=MainStreamProfileToken
    #  - rtsp://192.168.111.32:554/11
    pool_camera_sub:
      - onvif://admin:password1072@192.168.111.32:8000?subtype=SecondStreamProfileToken
    #  - rtsp://192.168.111.32:554/12
    driveway_camera:
      - onvif://hassio:Password123@192.168.111.30:8000?subtype=000
    driveway_camera_sub:
      - onvif://hassio:Password123@192.168.111.30:8000?subtype=001
    #  - rtsp://hassio:Password123@192.168.111.30:554/Preview_01_sub
    garage_cars_rot:
      - rtsp://192.168.111.31:554/ch0_0.h264
    garage_cars:
      - ffmpeg:garage_cars_rot#video=h264#rotate=180
#    living_room:
#      - rtsp://admin:password@192.168.111.8:554/play1.sdp

version: 0.15-0


cameras:
##########################
  driveway_camera:
    detect:
      width: 640
      height: 480
      fps: 5
    ffmpeg:
      hwaccel_args: preset-rpi-64-h264
      inputs:
##        - path: rtsp://frigate:silly40968192@192.168.111.30:554
#        - path: rtsp://hassio:Password123@192.168.111.30/Preview_01_main
        - path: rtsp://127.0.0.1:8554/driveway_camera
          input_args: preset-rtsp-restream
          roles:
#            - detect
#            - rtmp
            - record
        - path: rtsp://127.0.0.1:8554/driveway_camera_sub
          input_args: preset-rtsp-restream
          roles:
            - detect
#      user: hassio
#      password: Password123
    onvif:
      host: 192.168.111.30
      port: 8000
      user: hassio
      password: Password123

    zones:
      Driveway:
        coordinates: 0.002,0.341,0.845,0.162,0.997,0.291,0.995,0.952,0.002,0.97
        loitering_time: 0
    motion:
      mask: 0.001,0,0.999,-0.002,0.996,0.291,0.825,0.172,0.001,0.367
    review:
      alerts:
        required_zones: Driveway
      detections:
        required_zones: Driveway


  drivewaycars_camera:
    detect:
      width: 640
      height: 360
      fps: 5
    ffmpeg:
      hwaccel_args: preset-rpi-64-h264
      inputs:
        #- path: rtsp://admin:password@192.168.111.23:8554/Streaming/Channels/101
        - path: rtsp://127.0.0.1:8554/drivewaycars_camera
          input_args: preset-rtsp-restream
          roles:
            - record
#            - rtmp
#            - detect
        - path: rtsp://127.0.0.1:8554/drivewaycars_camera_sub
          input_args: preset-rtsp-restream
          roles:
            - detect
#      user: admin
#      password: silly40968192
#    audio:
#      enabled: true
    motion:
      mask: 
        0,0.293,0.002,0,0.999,0,0.999,0.781,0.885,0.456,0.7,0.424,0.701,0.311,0.507,0.294,0.447,0.313,0.377,0.293
    zones:
      Driveway:
        coordinates: 0.001,0.296,0.501,0.286,0.789,0.346,0.995,0.732,0.995,0.991,0.002,0.988
        loitering_time: 0
    review:
      alerts:
        required_zones: Driveway
      detections:
        required_zones: Driveway





detectors:
  hailo8l:
    type: hailo8l
#    device: M.2
    device: PCIe
    model:
      path: /config/model_cache/h8l_cache/ssd_mobilenet_v1.hef
model:
  width: 300
  height: 300
  input_tensor: nhwc
  input_pixel_format: bgr
  model_type: ssd
```

### docker-compose file or Docker CLI command

```yaml
HASSOS so whatever that installs
```

### Relevant Frigate log output

```shell
2024-11-15 19:29:19.198477518  [2024-11-15 14:29:19] detector.hailo8l               INFO    : Starting detection process: 397
2024-11-15 19:29:19.201025470  Process detector:hailo8l:
2024-11-15 19:29:19.201031433  Traceback (most recent call last):
2024-11-15 19:29:19.201032970    File "/opt/frigate/frigate/detectors/plugins/hailo8l.py", line 72, in __init__
2024-11-15 19:29:19.201033914      self.target = VDevice()
2024-11-15 19:29:19.201040081  NameError: name 'VDevice' is not defined
2024-11-15 19:29:19.201040729  
2024-11-15 19:29:19.201041785  During handling of the above exception, another exception occurred:
2024-11-15 19:29:19.201081340  
2024-11-15 19:29:19.201082840  Traceback (most recent call last):
2024-11-15 19:29:19.201084544    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2024-11-15 19:29:19.201085248      self.run()
2024-11-15 19:29:19.201086340    File "/opt/frigate/frigate/util/process.py", line 41, in run_wrapper
2024-11-15 19:29:19.201087174      return run(*args, **kwargs)
2024-11-15 19:29:19.201088248    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2024-11-15 19:29:19.201118174      self._target(*self._args, **self._kwargs)
2024-11-15 19:29:19.201119785    File "/opt/frigate/frigate/object_detection.py", line 109, in run_detector
2024-11-15 19:29:19.201133285      object_detector = LocalObjectDetector(detector_config=detector_config)
2024-11-15 19:29:19.201134415    File "/opt/frigate/frigate/object_detection.py", line 61, in __init__
2024-11-15 19:29:19.201135415      self.detect_api = create_detector(detector_config)
2024-11-15 19:29:19.201136489    File "/opt/frigate/frigate/detectors/__init__.py", line 18, in create_detector
2024-11-15 19:29:19.201137285      return api(detector_config)
2024-11-15 19:29:19.201138396    File "/opt/frigate/frigate/detectors/plugins/hailo8l.py", line 105, in __init__
2024-11-15 19:29:19.201139174      except HailoRTException as e:
2024-11-15 19:29:19.201153822  NameError: name 'HailoRTException' is not defined
2024-11-15 19:29:19.202797827  [2024-11-15 14:29:19] frigate.detectors.plugins.hailo8l INFO    : Initializing Hailo device as PCIe
2024-11-15 19:29:19.203105402  [2024-11-15 14:29:19] frigate.detectors.plugins.hailo8l INFO    : A model file already exists at /config/model_cache/h8l_cache/ssd_mobilenet_v1.hef not downloading one.
2024-11-15 19:29:19.228315456  [2024-11-15 14:29:19] frigate.app                    INFO    : Output process started: 413
```

### Install method

HassOS Addon

### Object Detector

Other

### Screenshots of the Frigate UI's System metrics pages

hailo0
Subsystem:
hailo_chardev
Device path:
/dev/hailo0
Attributes:
DEVNAME: /dev/hailo0
DEVPATH: /devices/virtual/hailo_chardev/hailo0
MAJOR: '511'
MINOR: '0'
SUBSYSTEM: hailo_chardev

### Any other information that may be helpful

I was hoping to get the Hailo8L from the AI Kit to work with Frigate so I can do motion sensing on a couple cameras.
Since I already have Home Assistant running on an M.2 SSD on my Raspberry Pi, I added the Hailo8L M.2 to the dual M.2 Hat I have. I could not try with the M.2 Hat that came with the AI Kit. Unfortunately, it gives the error above when it starts.
Home Assistant sees the Hailo according to the System Hardware page.

From my limited SW experience (I do embedded electronic HW not SW for a living), it seems there is a dependency missing.

**Comment by NickM-27** [comment]
please read the docs, HA OS does not support the hailo device currently, there would need to be a separate addon for that

**Comment by roblamoreaux** [comment]
Where is that in the docs, since I did not find it when I looked?

Is there plans to support it?

**Reply by NickM-27** [reply]
The docs explain that you need to specifically use the correct docker image, which the list of addons does not support. https://deploy-preview-13787--frigate-docs.netlify.app/frigate/installation#rockchip-platform

I do not know if support is planned, realistically it is outside of the scope of HA OS to have the user install the hailo driver, so I am not sure it will be officially supported

**Reply by roblamoreaux** [reply]
I had read that, but it was not clear that it was the only way, and being a beta I hoped that it was a work in progress,

Disappointing, but that is part of why it is a hobby not a critical function for me.

**Reply by foxgiggle** [reply]
I managed to run the Hailo8L in conjunction with the Frigate add-on for Home Assistant on the same hardware setup (RPI5 + dual M.2 PCIe + Hailo8L + NVMe). This can be done by using Debian (Raspbian) OS with Home Assistant Supervised instead of HA OS. This is a not recommended and relatively complex method of installing Home Assistant, but I couldn't find an alternative.
In Raspbian, you need to install the driver and firmware for the Hailo8L, after which the Frigate add-on can work with the Hailo8L.
P.S. The add-on's config.yaml needs to be modified to reference the dev version of Frigate with Hailo support, such as d12c780-h8l.

**Reply by roblamoreaux** [reply]
according to this issue for HAOS [#3680](https://github.com/home-assistant/operating-system/pull/3680) it looks like they are adding support for the Hailo8, so it should be possible in the next release (I think).

**Reply by sairon** [reply]
Yep, it should be possible to use Hailo-8 devices (meaning all current Raspberry Pi AI HATs and Kit) on Raspberry Pi 5 with HAOS 14.0.rc1 and later. I didn't have opportunity to test it with Frigate but from another report I have verified that the driver and firmware loads correctly now.

**Reply by NathanMTL** [reply]
If anyone has done this, dual nvme hat on rpi5 with hailo, could you share the inference times? 

I wanted to buy the dual hat for my frigate set up and hailo chip to also use nvme storage but I noticed they all seem to only support pcie 2 speeds. While testing my rpi5 and frigate 0.15 beta 1, I randomly started getting really slow inference times, like from 20ms to 50ms, I finally realized the pcie speed had reverted to 2 instead of 3, once I put I back, the inference returned.

Due to that, I decided not to buy the dual hat and just stick with my attached usb 3 ssd.

**Reply by foxgiggle** [reply]
I am using the Waveshare PCIe To 2-ch M.2 Adapter to connect the Hailo8L and PM991a NVMe to the RPI5 in PCIe Gen2 mode. The inference is in the range of 26-28 ms. When the Hailo8L was connected at PCIe Gen3 speed, the inference was in the range of 18-21 ms.
![image](https://github.com/user-attachments/assets/c494f58b-8868-4477-8d34-22c11be4c76e)

**Reply by NathanMTL** [reply]
Cool, thanks! That’s actually not bad. I might try that.

**Reply by lekrom** [reply]
Is the Hailo8L supported in HAOS yet? If so, how do I get it working? Any special steps? Installed frigate as add on in HAOS. Tried the hailo8l detector config as shown. Frigate no longer starts up. Any advice?

**Reply by NickM-27** [reply]
It is supported in HA OS but it is an older version, we must wait for the RPi kernel to update to a newer version of HailoRT for a Frigate addon to be supported

**Reply by lekrom** [reply]
Thanks for the reply @NickM-27.
Sounds like I should have bought a Coral instead.

**Reply by roblamoreaux** [reply]
The latest haos released (OS 14.1) updated Hailo8 support to 4.19.0, so it may work now. I moved my frigate to another Pi5 so I haven't tried it

**Reply by lekrom** [reply]
I have that version of HAOS installed.
The Hailo devuce appears to be detected. I can see it the system/hardware section. 
![Screenshot_20250112_090634_Home Assistant](https://github.com/user-attachments/assets/94001502-3e78-40ef-976f-e7fbf665f735)
Still cant get frigate running though.

**Reply by NickM-27** [reply]
Because you need the hailo variant of the docker image, there's a PR up to add an addon for this

**Reply by sairon** [reply]
@NickM-27 HAOS 14.0 had Hailo support only on Raspberry Pi 5 with its in-tree driver (4.18.0), [HAOS 14.1](https://github.com/home-assistant/operating-system/releases/tag/14.1) has the driver included also for x86 and we build it from original sources, so we have a little more control over the version provided. I'll try to keep in mind to have it synced with what Frigate provides (ideally ping me if it were to be updated on Frigate side first).

**Reply by NickM-27** [reply]
Sounds good. The 0.15 release is planned to go out with 4.19

**Reply by lekrom** [reply]
Hi @NickM-27,


> Because you need the hailo variant of the docker image, there's a PR up to add an addon for this

I assume you are talking about PR #205 in the hass-adons repo? I see you merged it recently. I assume the hailo variant is now available?

Regards,
Lekrom

**Reply by NickM-27** [reply]
Yes

**Reply by lekrom** [reply]
Works fantasticly, 6 cameras @1080p,10fps  each, Installed as add on in fresh supervised HA install on RPI5. CPU is barely awake. Thanks a million!

**Reply by iBobik** [reply]
> @lekrom add on in fresh supervised HA
„add-on“ is feature od the Home Assistant OS, not „supervised“ (currently deprecated method of Home Assistant installation).

So what type of installation you are using?

I am interested in HA OS with Frigate Add-on on RPi 5 with AI hat. Will it work with accelerated detector and video decoding?
